openwisp_controller/config/base/vpn.py

```diff
@@ -291,3 +559,38 @@ def _auto_create_cert(self, name, common_name):
         cert.save()
         self.cert = cert
         return cert
+
+    def _auto_wireguard(self):
+        """
+        Automatically generates private and public key for wireguard
+        """
+        if not self.vpn._is_backend_type('wireguard') or (
+            self.private_key and self.public_key
+        ):
+            return
+        self.private_key, self.public_key = crypto.generate_wireguard_keys()
+
+    def _auto_vxlan(self):
+        """
+        Automatically generates VNI for VXLAN
+        """
+        if not self.vpn._is_backend_type('vxlan') or self.vni:
+            return
+        last_tunnel = (
+            self._meta.model.objects.filter(vpn=self.vpn).order_by('vni').last()
+        )
+        if last_tunnel:
+            self.vni = last_tunnel.vni + 1
+        else:
+            self.vni = 1
+
+    def _auto_ip(self):
+        if not self.vpn.subnet:
+            return
+        self.ip = self.vpn.subnet.request_ip()
+        try:
+            self.ip.full_clean()
+            self.ip.save()
+        except Exception:
+            raise
+            logger.exception('Got exception while creating device ip')
```

Comment:
It has been already handled in openwisp-ipam. This will create an additional query. 

https://github.com/openwisp/openwisp-ipam/blob/03b6a28f7a6cdb1d3d2e3436c7593f3181be94e7/openwisp_ipam/base/models.py#L130-L141

Reply:
ok, we can remove this